File: linux/src/services/jsonbackup.cpp
#include "services/jsonbackup.h"

#include <QDateTime>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QSqlError>
#include <QSqlQuery>

#include <algorithm>

#include "data/database.h"
#include "data/goalrepository.h"
#include "data/taskrepository.h"
#include "model/enums.h"
#include "model/goal.h"
#include "model/taskitem.h"

namespace tick {

namespace {

QJsonValue optToString(const std::optional<std::string>& v) {
    return v.has_value() ? QJsonValue(QString::fromStdString(*v)) : QJsonValue(QJsonValue::Null);
}

QJsonValue optToDateTime(const std::optional<QDateTime>& v) {
    return v.has_value() ? QJsonValue(v->toString(Qt::ISODateWithMs)) : QJsonValue(QJsonValue::Null);
}

std::optional<QDateTime> dateTimeFromJson(const QJsonValue& v) {
    return dateTimeFromString(v.toString());
}

} // namespace

QJsonObject JsonBackup::taskToJson(const std::shared_ptr<TaskItem>& task) {
    QJsonObject o;
    o.insert(QStringLiteral("id"), QString::fromStdString(task->id));
    o.insert(QStringLiteral("name"), task->name);
    o.insert(QStringLiteral("colorHex"), optToString(task->colorHex));
    o.insert(QStringLiteral("iconSystemName"), optToString(task->iconSystemName));
    o.insert(QStringLiteral("typeRaw"), taskTypeToString(task->type));
    o.insert(QStringLiteral("statusRaw"), taskStatusToString(task->status));
    o.insert(QStringLiteral("totalAmount"), task->totalAmount);
    o.insert(QStringLiteral("currentAmount"), task->currentAmount);
    o.insert(QStringLiteral("startDate"), optToDateTime(task->startDate));
    o.insert(QStringLiteral("endDate"), optToDateTime(task->endDate));
    o.insert(QStringLiteral("reminderDate"), optToDateTime(task->reminderDate));
    o.insert(QStringLiteral("repeatRuleRaw"), task->repeatRule == RepeatRule::Never
                                                 ? QJsonValue(QJsonValue::Null)
                                                 : QJsonValue(repeatRuleToString(task->repeatRule)));
    o.insert(QStringLiteral("customWeekdaysRaw"), optToString(task->customWeekdaysRaw));
    o.insert(QStringLiteral("createdAt"), task->createdAt.toString(Qt::ISODateWithMs));
    o.insert(QStringLiteral("sortOrder"), task->sortOrder);
    QJsonArray subs;
    for (const auto& c : task->subtasks) {
        subs.append(taskToJson(c));
    }
    o.insert(QStringLiteral("subtasks"), subs);
    return o;
}

QJsonObject JsonBackup::goalToJson(const std::shared_ptr<Goal>& goal) {
    QJsonObject o;
    o.insert(QStringLiteral("id"), QString::fromStdString(goal->id));
    o.insert(QStringLiteral("name"), goal->name);
    o.insert(QStringLiteral("colorHex"), QString::fromStdString(goal->colorHex));
    o.insert(QStringLiteral("iconSystemName"), optToString(goal->iconSystemName));
    o.insert(QStringLiteral("startDate"), optToDateTime(goal->startDate));
    o.insert(QStringLiteral("endDate"), optToDateTime(goal->endDate));
    o.insert(QStringLiteral("startDatePreciseToHour"), goal->startDatePreciseToHour);
    o.insert(QStringLiteral("endDatePreciseToHour"), goal->endDatePreciseToHour);
    o.insert(QStringLiteral("createdAt"), goal->createdAt.toString(Qt::ISODateWithMs));
    o.insert(QStringLiteral("progressCountingModeRaw"), countingModeToString(goal->progressCountingMode));
    QJsonArray tasks;
    for (const auto& t : goal->tasks) {
        tasks.append(taskToJson(t));
    }
    o.insert(QStringLiteral("tasks"), tasks);
    return o;
}

QString JsonBackup::exportGoals() {
    GoalRepository goalRepo;
    TaskRepository taskRepo;
    const auto goals = goalRepo.all();
    QJsonArray arr;
    for (auto& g : goals) {
        g->tasks = taskRepo.loadForGoal(g->id);
        arr.append(goalToJson(g));
    }
    QJsonObject root;
    root.insert(QStringLiteral("version"), 1);
    root.insert(QStringLiteral("goals"), arr);
    return QString::fromUtf8(QJsonDocument(root).toJson(QJsonDocument::Indented));
}

std::shared_ptr<TaskItem> JsonBackup::taskFromJson(const QJsonValue& v) {
    auto task = std::make_shared<TaskItem>();
    if (v.isObject()) {
        const QJsonObject o = v.toObject();
        QString id = o.value(QStringLiteral("id")).toString();
        if (!id.isEmpty()) task->id = id.toStdString();
        task->name = o.value(QStringLiteral("name")).toString();
        const QString color = o.value(QStringLiteral("colorHex")).toString();
        task->colorHex = color.isEmpty() ? std::nullopt : std::optional<std::string>(color.toStdString());
        const QString icon = o.value(QStringLiteral("iconSystemName")).toString();
        task->iconSystemName = icon.isEmpty() ? std::nullopt : std::optional<std::string>(icon.toStdString());
        task->type = taskTypeFromString(o.value(QStringLiteral("typeRaw")).toString());
        task->status = taskStatusFromString(o.value(QStringLiteral("statusRaw")).toString());
        task->totalAmount = std::max(0.0, o.value(QStringLiteral("totalAmount")).toDouble());
        task->currentAmount = std::min(std::max(o.value(QStringLiteral("currentAmount")).toDouble(), 0.0),
                                       task->totalAmount);
        task->startDate = dateTimeFromJson(o.value(QStringLiteral("startDate")));
        task->endDate = dateTimeFromJson(o.value(QStringLiteral("endDate")));
        task->reminderDate = dateTimeFromJson(o.value(QStringLiteral("reminderDate")));
        task->repeatRule = repeatRuleFromString(o.value(QStringLiteral("repeatRuleRaw")).toString());
        const QString cw = o.value(QStringLiteral("customWeekdaysRaw")).toString();
        task->customWeekdaysRaw = cw.isEmpty() ? std::nullopt : std::optional<std::string>(cw.toStdString());
        const QDateTime created = dateTimeFromString(o.value(QStringLiteral("createdAt")).toString());
        if (created.isValid()) task->createdAt = created;
        task->sortOrder = o.value(QStringLiteral("sortOrder")).toInt();
        for (const auto& c : o.value(QStringLiteral("subtasks")).toArray()) {
            auto child = taskFromJson(c);
            child->parent = task;
            task->subtasks.push_back(child);
        }
    }
    return task;
}

std::shared_ptr<Goal> JsonBackup::goalFromJson(const QJsonObject& o) {
    auto goal = std::make_shared<Goal>();
    QString id = o.value(QStringLiteral("id")).toString();
    if (!id.isEmpty()) goal->id = id.toStdString();
    goal->name = o.value(QStringLiteral("name")).toString();
    const QString color = o.value(QStringLiteral("colorHex")).toString();
    if (!color.isEmpty()) goal->colorHex = color.toStdString();
    const QString icon = o.value(QStringLiteral("iconSystemName")).toString();
    goal->iconSystemName = icon.isEmpty() ? std::nullopt : std::optional<std::string>(icon.toStdString());
    goal->startDate = dateTimeFromJson(o.value(QStringLiteral("startDate")));
    goal->endDate = dateTimeFromJson(o.value(QStringLiteral("endDate")));
    goal->startDatePreciseToHour = o.value(QStringLiteral("startDatePreciseToHour")).toBool(false);
    goal->endDatePreciseToHour = o.value(QStringLiteral("endDatePreciseToHour")).toBool(false);
    const QDateTime created = dateTimeFromString(o.value(QStringLiteral("createdAt")).toString());
    if (created.isValid()) goal->createdAt = created;
    goal->progressCountingMode = countingModeFromString(
        o.value(QStringLiteral("progressCountingModeRaw")).toString());
    for (const auto& t : o.value(QStringLiteral("tasks")).toArray()) {
        goal->tasks.push_back(taskFromJson(t));
    }
    return goal;
}

bool JsonBackup::importGoals(const QString& json, QString& error) {
    const QJsonDocument doc = QJsonDocument::fromJson(json.toUtf8());
    if (!doc.isObject()) {
        error = QStringLiteral("备份文件格式错误（不是有效的 JSON 对象）");
        return false;
    }
    const QJsonObject root = doc.object();
    const QJsonArray goals = root.value(QStringLiteral("goals")).toArray();

    QSqlQuery clearQuery(Database::instance().connection());
    // 外键级联删除全部任务，仅供导入前清空
    if (!clearQuery.exec(QStringLiteral("DELETE FROM TaskItem"))) {
        error = QStringLiteral("清理旧数据失败：%1").arg(clearQuery.lastError().text());
        return false;
    }
    if (!clearQuery.exec(QStringLiteral("DELETE FROM Goal"))) {
        error = QStringLiteral("清理旧数据失败：%1").arg(clearQuery.lastError().text());
        return false;
    }

    GoalRepository goalRepo;
    TaskRepository taskRepo;
    for (const QJsonValue& gv : goals) {
        const auto goal = goalFromJson(gv.toObject());
        if (goal->name.isEmpty()) continue;
        if (!goalRepo.insert(*goal)) {
            error = QStringLiteral("恢复目标失败：%1").arg(goal->name);
            return false;
        }
        for (const auto& t : goal->tasks) {
            // save 会递归保存整棵子树并正确写入 goal_id / parent_task_id
            taskRepo.save(t, goal->id);
        }
    }
    return true;
}

} // namespace tick

File: windows/Services/JsonBackup.cs
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using Tick.Models;

namespace Tick.Services;

/// <summary>导出/导入备份页面选择文件后的 JSON 序列化器。</summary>
public sealed class JsonBackup
{
    /// <summary>JSON 序列化选项（宽松、保留缩进便于阅读）</summary>
    private static readonly JsonSerializerOptions Options = new()
    {
        WriteIndented = true,
    };

    /// <summary>导出全部目标（含任务树）为 JSON 字符串。</summary>
    public static string Serialize(IEnumerable<Goal> goals)
    {
        var doc = new BackupDocument { Goals = new List<GoalDto>() };
        foreach (var g in goals)
            doc.Goals.Add(GoalDto.From(g));
        return JsonSerializer.Serialize(doc, Options);
    }

    /// <summary>从 JSON 字符串还原目标列表（含任务树）。失败抛 <see cref="JsonException"/>。</summary>
    public static List<Goal> DeserializeIntoGoals(string json)
    {
        var doc = JsonSerializer.Deserialize<BackupDocument>(json, Options)
                  ?? throw new JsonException("备份文件格式无效");
        var result = new List<Goal>();
        foreach (var dto in doc.Goals ?? new List<GoalDto>())
            result.Add(dto.ToGoal());
        return result;
    }

    // ---- 备份文档模型 ----

    private sealed class BackupDocument
    {
        public string App { get; set; } = "Tick";
        public int Version { get; set; } = 1;
        public List<GoalDto>? Goals { get; set; }
    }

    private sealed class GoalDto
    {
        public Guid Id { get; set; }
        public string Name { get; set; } = "";
        public string ColorHex { get; set; } = "auto";
        public string? IconSystemName { get; set; }
        public DateTime? StartDate { get; set; }
        public DateTime? EndDate { get; set; }
        public bool StartDatePreciseToHour { get; set; }
        public bool EndDatePreciseToHour { get; set; }
        public DateTime CreatedAt { get; set; }
        public string ProgressCountingModeRaw { get; set; } = ProgressCountingMode.AllTasks.ToString();
        public List<TaskDto>? Tasks { get; set; }

        public static GoalDto From(Goal g) => new()
        {
            Id = g.Id,
            Name = g.Name,
            ColorHex = g.ColorHex,
            IconSystemName = g.IconSystemName,
            StartDate = g.StartDate,
            EndDate = g.EndDate,
            StartDatePreciseToHour = g.StartDatePreciseToHour,
            EndDatePreciseToHour = g.EndDatePreciseToHour,
            CreatedAt = g.CreatedAt,
            ProgressCountingModeRaw = g.ProgressCountingModeRaw,
            Tasks = TaskDto.FromList(g.Tasks),
        };

        public Goal ToGoal()
        {
            var goal = new Goal
            {
                Id = Id,
                Name = Name,
                ColorHex = ColorHex,
                IconSystemName = IconSystemName,
                StartDate = StartDate,
                EndDate = EndDate,
                StartDatePreciseToHour = StartDatePreciseToHour,
                EndDatePreciseToHour = EndDatePreciseToHour,
                CreatedAt = CreatedAt,
                ProgressCountingModeRaw = ProgressCountingModeRaw,
            };
            goal.Tasks.AddRange(TaskDto.TreeOf(Tasks));
            foreach (var task in goal.Tasks)
                task.Goal = goal;
            return goal;
        }
    }

    private sealed class TaskDto
    {
        public Guid Id { get; set; }
        public string Name { get; set; } = "";
        public string? ColorHex { get; set; }
        public string? IconSystemName { get; set; }
        public string TypeRaw { get; set; } = TaskType.Single.ToString();
        public string StatusRaw { get; set; } = TaskStatus.NotDone.ToString();
        public double TotalAmount { get; set; }
        public double CurrentAmount { get; set; }
        public DateTime? StartDate { get; set; }
        public DateTime? EndDate { get; set; }
        public DateTime? ReminderDate { get; set; }
        public string? RepeatRuleRaw { get; set; }
        public string? CustomWeekdaysRaw { get; set; }
        public DateTime CreatedAt { get; set; }
        public int SortOrder { get; set; }
        public List<TaskDto>? Subtasks { get; set; }

        public static List<TaskDto> FromList(IEnumerable<TaskItem> tasks)
        {
            var list = new List<TaskDto>();
            foreach (var t in tasks)
                list.Add(From(t));
            return list;
        }

        private static TaskDto From(TaskItem t) => new()
        {
            Id = t.Id,
            Name = t.Name,
            ColorHex = t.ColorHex,
            IconSystemName = t.IconSystemName,
            TypeRaw = t.TypeRaw,
            StatusRaw = t.StatusRaw,
            TotalAmount = t.TotalAmount,
            CurrentAmount = t.CurrentAmount,
            StartDate = t.StartDate,
            EndDate = t.EndDate,
            ReminderDate = t.ReminderDate,
            RepeatRuleRaw = t.RepeatRuleRaw,
            CustomWeekdaysRaw = t.CustomWeekdaysRaw,
            CreatedAt = t.CreatedAt,
            SortOrder = t.SortOrder,
            Subtasks = t.Subtasks.Count > 0 ? FromList(t.Subtasks) : null,
        };

        /// <summary>把 DTO 树还原为 TaskItem 对象图（重建父链，Goal 由外层填充）。</summary>
        public static List<TaskItem> TreeOf(List<TaskDto>? dtos)
        {
            var result = new List<TaskItem>();
            if (dtos is null)
                return result;
            foreach (var dto in dtos)
                result.Add(ToTask(dto));
            Link(result);
            return result;
        }

        private static TaskItem ToTask(TaskDto d) => new()
        {
            Id = d.Id,
            Name = d.Name,
            ColorHex = d.ColorHex,
            IconSystemName = d.IconSystemName,
            TypeRaw = d.TypeRaw,
            StatusRaw = d.StatusRaw,
            TotalAmount = d.TotalAmount,
            CurrentAmount = d.CurrentAmount,
            StartDate = d.StartDate,
            EndDate = d.EndDate,
            ReminderDate = d.ReminderDate,
            RepeatRuleRaw = d.RepeatRuleRaw,
            CustomWeekdaysRaw = d.CustomWeekdaysRaw,
            CreatedAt = d.CreatedAt,
            SortOrder = d.SortOrder,
        };

        /// <summary>单遍关联父子（DTO 有序，父在前，子已展开为当前根列表的子树）。</summary>
        private static void Link(List<TaskItem> roots)
        {
            var stack = new Stack<TaskItem>();
            foreach (var root in roots)
            {
                LinkSubtree(root, stack);
            }
            foreach (var root in roots)
            {
                if (root.ParentTask is not null)
                    root.ParentTask.Subtasks.Add(root);
            }
        }

        private static void LinkSubtree(TaskItem item, Stack<TaskItem> stack)
        {
            foreach (var maybeChild in item.Subtasks)
            {
                maybeChild.ParentTask = item;
                LinkSubtree(maybeChild, stack);
            }
        }
    }
}